### PR TITLE
Ensure exported struct in `sdk/log` are not comparable

### PR DESCRIPTION
**Pull request by MrAlias**
Ensure the `Record` and `Provider` continue to be non-comparable.

Restrict `BatchProcessor` and `SimpleProcessor` to be non-comparable.

There is no obvious reason an end-user will need to compare these types, and we want to keep the possibility of changing the internals without changing this behavior.

if comparability is required by end-users in the future we can add `Equal(other T)` methods in the future.

### CODE REVIEW

File: sdk/log/batch.go [hunk 1]
```diff
@@ -96,6 +96,8 @@ type BatchProcessor struct {
 
 	// stopped holds the stopped state of the BatchProcessor.
 	stopped atomic.Bool
+
+	noCmp [0]func() //nolint: unused  // This is indeed used.
```
Comment (XSAM) [inline sdk/log/batch.go:100]:
```suggestion
	noCmp [0]func() //nolint: unused  // This is indeed used to make this struct non-comparable.
```

Nit. To make it clear.

File: sdk/log/provider.go [hunk 2]
```diff
@@ -69,6 +69,8 @@ type LoggerProvider struct {
 	loggers   map[instrumentation.Scope]*logger
 
 	stopped atomic.Bool
+
+	noCmp [0]func() //nolint: unused  // This is indeed used.
```
Comment (XSAM) [inline sdk/log/provider.go:73]:
```suggestion
	noCmp [0]func() //nolint: unused  // This is indeed used to make this struct non-comparable.
```

File: sdk/log/record.go [hunk 3]
```diff
@@ -86,6 +86,8 @@ type Record struct {
 
 	attributeValueLengthLimit int
 	attributeCountLimit       int
+
+	noCmp [0]func() //nolint: unused  // This is indeed used.
```
Comment (XSAM) [inline sdk/log/record.go:90]:
```suggestion
	noCmp [0]func() //nolint: unused  // This is indeed used to make this struct non-comparable.
```

File: sdk/log/simple.go [hunk 4]
```diff
@@ -16,6 +16,8 @@ var _ Processor = (*SimpleProcessor)(nil)
 // Use [NewSimpleProcessor] to create a SimpleProcessor.
 type SimpleProcessor struct {
 	exporter Exporter
+
+	noCmp [0]func() //nolint: unused  // This is indeed used.
```
Comment (XSAM) [inline sdk/log/simple.go:20]:
```suggestion
	noCmp [0]func() //nolint: unused  // This is indeed used to make this struct non-comparable.
```

**Review by XSAM — APPROVED** [review]

File: sdk/log/record.go (hunk 3 above)
Comment (pellared) [inline sdk/log/record.go:90]: This is not necessary, but may be good for maintaining backward compatiblity.
Should we do the same for `Record`, `KeyValue`, Value` in `go.opentelemetry.io/otel/log`?

File: sdk/log/batch.go (hunk 1 above)
Comment (MrAlias) [reply sdk/log/batch.go:100]: I'm not in favor of that change. It makes the line too long, is verbose without adding addition information. From the context of the field type it is already implied that this is making the struct non-comparable.

This is a sentence we have used many times before, and continues to be appropriate at this point.

File: sdk/log/record.go (hunk 3 above)
Comment (MrAlias) [reply sdk/log/record.go:90]: > Should we do the same for `Record`, `KeyValue`, `Value` in `go.opentelemetry.io/otel/log`?

SGTM :+1:. Please feel free to open a PR.

**Review by dashpole — APPROVED** [review]

File: sdk/log/record.go (hunk 3 above)
Comment (pellared) [reply sdk/log/record.go:90]: I created https://github.com/open-telemetry/opentelemetry-go/issues/5696

**Review by pellared — APPROVED** [review]

**Pull request merged**